File: SmartBat/BTTimeToBeatTransmitter.m
//
//  BTTimeToBeatTransmitter.m
//  SmartBat
//
//  Created by poppy on 13-6-6.
//  Copyright (c) 2013年 kaka'. All rights reserved.
//

#import "BTTimeToBeatTransmitter.h"

#define SECONDS_PER_MINUTE 60;
#define DEFAULT_DURATION 0.5;

@implementation BTTimeToBeatTransmitter

-(id)init
{
    self = [super init];
    
    _globals = [BTGlobals sharedGlobals];
    _beatCount = 0;
    _minDistance = 100000000.0;
    _maxDistance =0;
    _avarageDistance = 0;
    _totalDistance = 0;
    
    return self;
}


//bpm
-(void) updateBPM:(int) bpm
{
    _bpm = bpm;
    
    _noteDuration = [self getNoteDurationByBPM:bpm andNote:_measureTemplate.noteType andSubdivision:_subdivisionTemplate];
    
    [self updateClockDuration];
}



//measure template
-(void) updateMeasureTemplate:(BTMeasure *) measureTemplate
{

    _noteDuration = [self getNoteDurationByBPM:_bpm andNote:measureTemplate.noteType andSubdivision:_subdivisionTemplate];
       

    _measureTemplate = measureTemplate;
     [self updateClockDuration];
}


-(void) updateSubdivisionTemplate:(BTSubdivision *)subdivisionTemplate
{

    _noteDuration = [self getNoteDurationByBPM:_bpm andNote:_measureTemplate.noteType andSubdivision:subdivisionTemplate];
        
    _subdivisionTemplate = subdivisionTemplate;
    [self updateClockDuration];
}


-(BTMeasure *)getMeasureTemplate
{
    return _measureTemplate;
}


-(void)updateClockDuration
{
    if(_globals.currentSubdivisionDuration!=_noteDuration)
    {
        _globals.currentSubdivisionDuration = _noteDuration;
    }
    
    
    double currentNoteDuration = _noteDuration * [_subdivisionTemplate count];
    if(_globals.currentNoteDuration != currentNoteDuration)
    {
        _globals.currentNoteDuration = currentNoteDuration;
    }
    
    
    double currentMeasureDuration = _noteDuration * [_subdivisionTemplate count] * [_measureTemplate getNoteCount ];
    if(_globals.currentMeasureDuration != currentMeasureDuration)
    {
        _globals.currentMeasureDuration = currentMeasureDuration;
    }
    
    [_timeLine updateClockDuration:_noteDuration];
}




-(void) bindTimeLine:(BTTimeLine *) timeLine
{
    _timeLine = timeLine;
    _timeLine.timeLineDelegate = self;
}

-(double) startWithBPM:(int)bpm andMeasureTemplate:(BTMeasure *)measureTemplate andSubdivision:(BTSubdivision *) subdivision
{
    
    _bpm = bpm;
    _subdivisionTemplate = subdivision;
    _noteDuration = [self getNoteDurationByBPM:bpm andNote:measureTemplate.noteType andSubdivision:_subdivisionTemplate];
    _measureTemplate = measureTemplate;
    
    
    if(_timeLine)
    {
        
        [self updateClockDuration];
        return [_timeLine startLoopWithDuration:_noteDuration];
    }
    
    return -1;
}

-(double)start
{
    if(_timeLine)
    {
        return [_timeLine startLoopWithDuration:_noteDuration];
    }
    return -1;
}

-(double)stop
{
    if(_timeLine)
    {
        
        [_measureTemplate reset];
        [_subdivisionTemplate reset];
        return [_timeLine stopLoop];
    }
    return -1;
}

-(double)getNoteDurationByBPM:(int) bpm andNote:(double)note andSubdivision:(BTSubdivision *)subdivision
{
    double duration = 60.0/(bpm/(note*4)* [subdivision count]);
    return duration;
}

-(double) getNowTime
{
    return [[NSDate date] timeIntervalSince1970];
}

//implement of TimeLineDelegate
-(void)onTimeInvokeHandler: (double) time
{
    
    _beatCount ++;
    
    BTBeat * beat = [_measureTemplate getCurrentNote];
    beat.indexOfMeasure = _measureTemplate.playIndex;
    beat.indexOfSubdivision = _subdivisionTemplate.playIndex;
    beat.hitTime = time;

    
    switch(_subdivisionTemplate.playIndex)
    {
        case 0:
            [self.timeToBeatTransmitterBeatDelegate onBeatHandler:beat ofMeasure:_measureTemplate withBPM:_bpm];
            
            [_measureTemplate playNote];
            [_subdivisionTemplate playNote ];
            break;
        default:
            [self.timeToBeatTransmitterBeatDelegate onSubdivisionHandler:beat];
            
            [_subdivisionTemplate playNote ];
            break;
    }

}



@end

File: java/src/com/sensormusic/metronomecore/BTTimeToBeatTransmitter.java
package com.sensormusic.metronomecore;



public class BTTimeToBeatTransmitter implements BTTimeLineDelegate {

	private double _noteDuration;
	private float _BPM = -1;

	private BTMeasure _measureTemplate;
	private BTSubdivision _subdivisionTemplate;

	private BTTimeLine _timeLine;
	private BTTimeToBeatTransmitterDelegate _transmitterDelegate;

	public BTTimeToBeatTransmitter() {
		
	}

	/**
	 * implement BTTimeLineDelegate
	 */
	public void onTimeInvokeHandler(double time) {

		switch (this._subdivisionTemplate.playIndex) {		
		case 0:
			
			BTBeat beat = this._measureTemplate.getCurrentNote();
			beat.indexOfMeasure = this._measureTemplate.playIndex;
			beat.indexOfSubdivision = this._subdivisionTemplate.playIndex;
			beat.hitTime = time;
			
			if(this._transmitterDelegate!= null){
				this._transmitterDelegate.onNoteHitHandler(beat);
			}
			
			
//			System.out.printf("%d @ %f\n",beat.beatType.getTypeNum(), time);
			this._measureTemplate.playNote();
			this._subdivisionTemplate.playNote();
			break;
			
		default:
			
			BTBeat subdivisionBeat = this._subdivisionTemplate.getCurrentNote();
			subdivisionBeat.indexOfMeasure = this._measureTemplate.playIndex;
			subdivisionBeat.indexOfSubdivision = this._subdivisionTemplate.playIndex;
			subdivisionBeat.hitTime = time;
			
			if(this._transmitterDelegate!= null){
				this._transmitterDelegate.onSubdivisionHitHandler(subdivisionBeat);
			}
			
//			System.out.printf("%d @ %f [s]\n",subdivisionBeat.beatType.getTypeNum(), time);
			this._subdivisionTemplate.playNote();
			break;
		}

	}

	
	
	public void updateBPM(float BPM) {

		this._BPM = BPM;

		this._noteDuration = calculateNoteDuration(this._BPM,
				this._measureTemplate.noteType, this._subdivisionTemplate);

		updateClockDuration();
	}

	
	
	public void updateMeasureTemplate(BTMeasure measureTemplate) {

		

		this._noteDuration = calculateNoteDuration(this._BPM,
				measureTemplate.noteType, this._subdivisionTemplate);
		
		this._measureTemplate = measureTemplate;

		updateClockDuration();
	}

	
	
	public void updateSubdivisionTemplate(BTSubdivision subdivisionTemplate) {
		
		

		this._noteDuration = calculateNoteDuration(this._BPM,
				this._measureTemplate.noteType, subdivisionTemplate);
		
		this._subdivisionTemplate = subdivisionTemplate;

		updateClockDuration();
	}
	

	public void bindTimeLine(BTTimeLine timeLine) {

		this._timeLine = timeLine;
		this._timeLine.setTimeLineDelegate(this);

	}

	public double start(float BPM, BTMeasure measureTemplate,
			BTSubdivision subdivisionTemplate) {

		this._BPM = BPM;
		this._subdivisionTemplate = subdivisionTemplate;
		this._noteDuration = calculateNoteDuration(BPM,
				measureTemplate.noteType, subdivisionTemplate);
		this._measureTemplate = measureTemplate;

		if (this._timeLine != null) {

			updateClockDuration();
			return this._timeLine.startLoopWithDuration(this._noteDuration);
		}

		return -1;
	}

	public double start() {
		
	    if(this._timeLine!= null)
	    {
	        return this._timeLine.startLoopWithDuration(this._noteDuration);
	    }
	    return -1;
	}

	public double stop() {
		
		if(this._timeLine != null){
			this._measureTemplate.reset();
			this._subdivisionTemplate.reset();
			return this._timeLine.stopLoop();
		}
		
		return -1;
		
	}

	private double calculateNoteDuration(float BPM, BTNoteType noteType,
			BTSubdivision subdivisionTemplate) {
		
		if(BPM == -1 || noteType == null || subdivisionTemplate == null){
			return -1;
		}
		
	    double duration = 60.0/(BPM/( noteType.getTypeNum() *4) * subdivisionTemplate.count());
	    return duration;
	    
	}

	private void updateClockDuration()  {

		if(this._noteDuration < 0){
			
			return ;

		}
		this._timeLine.updateClockDuration(this._noteDuration);

	}


	public BTTimeToBeatTransmitterDelegate getTransmitterDelegate() {
		return _transmitterDelegate;
	}


	public void setTransmitterDelegate(BTTimeToBeatTransmitterDelegate _transmitterDelegate) {
		this._transmitterDelegate = _transmitterDelegate;
	}
}
